### PR TITLE
Reinstate internal adaptive and tuning

### PR DESCRIPTION
**Pull request by yallup**
Intermediate WIP PR where we reorganise things again. Compared to external-params this moves the external pieces back into the adaptive.py setup.

I had to force push this due to lots of mypy warnings in utils so I will address those as part of this (the utils need some addressing anyway, in particular finalise)

### CODE REVIEW

**Comment by yallup** [comment]
Can’t replicate issues with compile time so I’m going to say lgtm, ready to merge

File: blackjax/mcmc/ss.py [hunk 1]
```diff
@@ -335,8 +335,8 @@ def sample_direction_from_covariance(
     u = jax.random.normal(rng_key, shape=p.shape, dtype=p.dtype)
     u /= jnp.linalg.norm(u)
     L = jnp.linalg.cholesky(cov).astype(p.dtype)
-    dim = cov.shape[0]
-    L = L * 2 * jnp.sqrt(dim + 2)
+    # dim = cov.shape[0]
+    L = L * 2
```
Comment (AdamOrmondroyd) [inline blackjax/mcmc/ss.py]: this change is causing test_default_direction_generation() to fail

**Review by AdamOrmondroyd — CHANGES_REQUESTED** [review]

File: blackjax/mcmc/ss.py (hunk 1 above)
Comment (AdamOrmondroyd) [reply blackjax/mcmc/ss.py]: @yallup has convinced me this is a later problem

**Review by AdamOrmondroyd — APPROVED** [review]
lgtm

**Comment by yallup** [comment]
Compile time increase seems linked to https://github.com/jax-ml/jax/issues/21313

reverted to use previous linalg.inv and it seems to have fixed things.

File: blackjax/mcmc/ss.py [hunk 2]
```diff
@@ -332,12 +317,10 @@ def sample_direction_from_covariance(
...
-    u = jax.random.normal(rng_key, shape=p.shape, dtype=p.dtype)
-    u /= jnp.linalg.norm(u)
-    L = jnp.linalg.cholesky(cov).astype(p.dtype)
-    dim = cov.shape[0]
-    L = L * 2 * jnp.sqrt(dim + 2)
-    d = linear_map(L, u)
+    d = jax.random.normal(rng_key, shape=p.shape, dtype=p.dtype)
+    invcov = jnp.linalg.inv(cov)
+    norm = jnp.sqrt(jnp.einsum("...i,...ij,...j", d, invcov, d))
+    d = d / norm[..., None]
```
Comment (williamjameshandley) [inline blackjax/mcmc/ss.py:323]: Here we had some conversation with @zwei-beiner about how in fact it is correct to normalise this way (non-trivially). I'm happy that this is now the correct d scaling, do we want a factor of 2 (diameter rather than radius) or something similar?

File: blackjax/ns/adaptive.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    """
+
+    def kernel(
+        rng_key: PRNGKey, state: AdaptiveNSState
+    ) -> tuple[AdaptiveNSState, NSInfo]:
+        """Performs one step of adaptive Nested Sampling."""
+        adapted_kernel = base_build_kernel(
+            delete_fn,
+            partial(inner_kernel, **state.inner_kernel_params),
+        )
```
Comment (williamjameshandley) [inline blackjax/ns/adaptive.py:97]: This is so much neater.

File: blackjax/ns/nss.py [hunk 4]
```diff
@@ -230,9 +222,10 @@ def init_fn(position, rng_key=None):
         return init(
             position,
             init_state_fn=jax.vmap(init_state_fn),
+            update_inner_kernel_params_fn=update_inner_kernel_params_fn,
         )
 
-    def step_fn(rng_key, state, inner_kernel_params):
-        return kernel(rng_key, state, inner_kernel_params)
+    def step_fn(rng_key, state):
+        return kernel(rng_key, state)
```
Comment (williamjameshandley) [inline blackjax/ns/nss.py:229]: this could be one line `step_fn = kernel`?

**Comment by williamjameshandley** [comment]
@codex review

File: blackjax/mcmc/ss.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (williamjameshandley) [reply blackjax/mcmc/ss.py:323]: Up to a factor of 2*sqrt(d+2) these are the same. @zwei-beiner showed that the factor should in fact be sqrt((d+2)/d) [non-trivially], so just dropping the d-dependence is best, hence only the factor of 2 remains important..
